Project2015To2017/ProjectConverter.cs

```diff
@@ -172,37 +188,114 @@ internal static bool Validate(FileInfo file, ILogger logger)
 			return false;
 		}
 
-		private IReadOnlyCollection<ITransformation> TransformationsToApply(bool modernProject)
+		private IReadOnlyCollection<ITransformation> TransformationsToApply()
+		{
+			var all = this.transformationSet.Transformations(this.logger, this.conversionOptions);
+			var (normal, others) = all.Split(FilterTargetNormalTransformations);
+			var (early, late) = others.Split(x =>
+				((ITransformationWithTargetMoment) x).ExecutionMoment == TargetTransformationExecutionMoment.Early);
+			var res = new List<ITransformation>(all.Count);
+			TopologicalSort(early, res, this.logger);
+			TopologicalSort(normal, res, this.logger);
+			TopologicalSort(late, res, this.logger);
+			return res;
+		}
+
+		private static void TopologicalSort(
+			IReadOnlyList<ITransformation> source,
+			ICollection<ITransformation> target,
+			ILogger logger)
+		{
+			var count = source.Count;
+			if (count == 0)
+			{
+				return;
+			}
+
+			// When Span<T> becomes available - replace with
+			// var used = count <= 256 ? stackalloc byte[count] : new byte[count];
+			var used = new byte[count];
+			var res = new LinkedList<ITransformation>();
+			var mappings = new Dictionary<string, int>();
+			for (var i = 0; i < count; i++)
+			{
+				var transformation = source[i];
+				if (transformation == null)
+				{
+					throw new ArgumentNullException(nameof(transformation),
+						"Transformation set must not contain null items");
+				}
```

Comment:
I feel that you favour performance over readability too much for the needs of this project. To me, this check and addition of the items to the list would be much clearer as an `Any()` call and a `Select().ToDictionary()` call. In practise nobody is trying to convert 10,000 projects and even with that many the difference would probably be minimal.

Reply:
I do use Linq methods, but in cases where I believe things will get simplified. This `for` loop is natural, its designation is obvious from looking at the code. If I replace that with 2 lines of Linq garbage — from my point of view it will get more complicated to grasp the meaning (I "love" `ToDictionary" arguments — who could've imagined less obvious API?). It's not performance optimization, rather it is just as I've been step-by-step writing this code. When I've been reviewing this code myself I didn't feel the urge to replace things with Linq.